samples/csharp_dotnetcore/12.customQABot/Bots/CustomQABot.cs

```diff
@@ -0,0 +1,133 @@
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.AI.QnA;
+using Microsoft.Bot.Builder.AI.QnA.Models;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.BotBuilderSamples
+{
+    public class CustomQABot : ActivityHandler
+    {
+        private readonly ILogger<CustomQABot> _logger;
+        private readonly IHttpClientFactory _httpClientFactory;
+        private readonly string _endpointKey;
+        private readonly string _hostname;
+        private readonly string _knowledgeBaseId;
+        private readonly string _defaultWelcome = "Hello and Welcome";
+        private readonly bool _enablePreciseAnswer = true;
+        private readonly bool _displayPreciseAnswerOnly = false;
+
+        public CustomQABot(IHttpClientFactory httpClientFactory, IConfiguration configuration, ILogger<CustomQABot> logger)
+        {
+            _logger = logger;
+            _httpClientFactory = httpClientFactory;
+
+            _hostname = configuration["LanguageEndpointHostName"];
+            if (string.IsNullOrEmpty(_hostname))
+            {
+                throw new ArgumentException(nameof(_hostname));
+            }
+
+            _endpointKey = configuration["LanguageEndpointKey"];
+            if (string.IsNullOrEmpty(_endpointKey))
+            {
+                throw new ArgumentException(nameof(_endpointKey));
+            }
+
+            _knowledgeBaseId = configuration["ProjectName"];
+            if (string.IsNullOrEmpty(_knowledgeBaseId))
+            {
+                throw new ArgumentException(nameof(_knowledgeBaseId));
+            }
+
+            var welcomeMsg = configuration["DefaultWelcomeMessage"];
+            if (!string.IsNullOrWhiteSpace(welcomeMsg))
+            {
+                _defaultWelcome = welcomeMsg;
+            }
+
+            _enablePreciseAnswer = !bool.TryParse(configuration["EnablePreciseAnswer"], out var enablePreciseAnswer) || enablePreciseAnswer;
+            _displayPreciseAnswerOnly = !bool.TryParse(configuration["DisplayPreciseAnswerOnly"], out var displayPreciseAnswerOnly) || displayPreciseAnswerOnly;
+        }
+
+        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
+        {
+
+            using var httpClient = _httpClientFactory.CreateClient();
+
+            var customQuestionAnswering = CreateCustomQuestionAnsweringClient(httpClient);
+
+            // Call Custom Question Answering service to get a response.
+            _logger.LogInformation("Calling Custom Question Answering");
+            var options = new QnAMakerOptions { Top = 1, EnablePreciseAnswer = _enablePreciseAnswer };
+            var response = await customQuestionAnswering.GetAnswersAsync(turnContext, options);
+
+            if (response?.Length > 0)
+            {
+                var activities = new List<Activity>();
```

Comment:
Make it list of IActivity to avoid potential issues when calling activites.ToArray() in line 100 below
```suggestion
                var activities = new List<IActivity>();
```